README.md
# Steering Wheel Interface

Android application for interfacing with the steering wheel radio control buttons in certain Jeep/Chrysler/Dodge vehicles.

## Purpose

This project exists primarily as an example of how to interface an Android device with a vehicle bus using an inexpensive ELM327 based OBDII scantool.

This is NOT an OBDII or "code reader" type application! A scantool is used only to provide the interface to other vehicle buses that happen to also use OBDII standard protocols (SAE J1850 PWM, SAE J1850 VPW, ISO 9141-2, ISO 14230-4 KWP, ISO 15765-4 CAN, SAE J1939 CAN).

The application is not likely useful to many people exactly as-is, but is designed to be extended to work with your specific vehicle/scenario. Note that only USB scantools are currently supported, though adding BlueTooth should be relatively simple.

## Getting Started

DISCLAIMER: EVERYTHING INCLUDED IN THIS REPOSITORY IS FOR INFORMATION PURPOSES ONLY. I AM NOT RESPONSIBLE FOR ANY DAMAGE TO YOU, YOUR VEHICLE, AND/OR DEVICES. DO NOT ATTEMPT ANY OF THE FOLLOWING UNLESS YOU KNOW WHAT YOU ARE DOING.

### To test the application as-is:

Requirements:

+	An Android device with USB Host support running Android 4.1 or newer.
	
	NOTE: The majority of the functionality only requires Android 3.1. If you remove or refactor the "settings screen" functionality, you should be able to compile and deploy to Honeycomb devices.
	
	Root access is required for some functionality.

+	An ELM327 based USB scantool
	
	NOTE: Many of the cheap eBay clones do not implement every protocol correctly and therefore may not work. [This one](http://www.amazon.com/ScanTool-423001-ElmScan-Diagnostic-Software/dp/B002PYBZJO/) is known to work.

+	A USB OTG cable	(or an Android device with native USB Host port)

+	A Jeep/Chrysler/Dodge vehicle with a SAE J1850 VPW base PCI Bus (many late 90's through mid 2000 models)

Compile and deploy the application. Plug the scantool into the vehicle via the OBDII diagnostic port and into the Android device through the USB OTG cable. 

Respond to the Android prompt regarding a USB device being attached by selecting this app to launch. It is recommended you select the "always" option. NOTE: If you already have a default application for USB-Serial devices, you will need to clear that setting to ever see this prompt.

A notification should be seen saying the Steering Wheel Interface started. Click the notification for the settings screen where you can adjust the baud rate for your particular scantool. IMPORTANT: You will need to restart the app for the new baud rate (and some other settings) to take effect!

Press your steering wheel radio control buttons. By default, play/pause, next/prev track, volume up/down, and go-to-home-screen actions are supported. IMPORTANT: Only the volume actions do not require root, the other actions will fail silently on non-rooted devices!

Tested and confirmed working on a Motorola XOOM 4G running the CyanogenMod 10 ROM (Android 4.1) with a 2003 Jeep Grand Cherokee. UPDATE: Confirmed working in CM 10.1 (Android 4.2) as well.

### Using with other vehicle years/makes/models:

+	Step 1: Determine how to interface (via USB scantool) with the "comfort bus" or similar in your vehicle. This may involve modifications to the scantool and/or vehicle wiring.

	NOTE: Many older vehicles have their buses interconnected, thereby allowing you to simply connect to the OBDII port. Most newer vehicles isolate the Diagnostic Bus (on the OBDII port). For those vehicles you would have to splice into another bus directly. Extra pins on the OBDII port may expose additional buses. Also the radio harness typically exposes the comfort bus. Get a factory service manual with wiring diagrams or Google and hope someone else has hacked on your specific vehicle year/make/model.

	See [this article](http://theksmith.com/technology/hack-vehicle-bus-cheap-easy-part-1/) for more info.

+	Step 2: Determine what bus messages are sent when interacting with the factory device you wish to monitor. See [this article](http://theksmith.com/technology/hack-vehicle-bus-cheap-easy-part-2/) for an example how-to.

+	Step 3: Adjust the app as needed based on your findings. In particular, the BUTTON_* declarations in ButtonActions.java, the "Scantool Protocol" in settings, and the "Scantool Monitor Command".

	If all messages that you wish to monitor for are sent from or to a particular device id, use ATMT## or ATMR## for the monitor command. See the ELM327 datasheet for more info. You could use ATMA instead to monitor all bus messages, though this could affect performance.

## Copyright and License

Copyright 2013 Kristoffer Smith and other contributors.

Except for external references, this project is licensed under the MIT License.

[http://opensource.org/licenses/MIT](http://opensource.org/licenses/MIT)

## Contributing

Pull requests are welcome! Please contact me if you'd like to do a major re-work or extension of the project.

[https://github.com/theksmith/Steering-Wheel-Interface](https://github.com/theksmith/Steering-Wheel-Interface)

## Credits

### Authors

+	Kristoffer Smith ([http://theksmith.com/](http://theksmith.com/))

### Contributors

+	none yet

### References & Dependencies

+	This project uses the usb-serial-for-android library. Source code is included instead of the binary as some of the latest unreleased features were needed. The source has not been modified for this release. Note that this library is licensed under LGPL Version 2.1.

	[https://github.com/mik3y/usb-serial-for-android](https://github.com/mik3y/usb-serial-for-android)

CHANGELOG.md
# Version 1.2 (9/18/2013)

+	Minor bug fixes and general code cleanup
+	Improved documentation, including new basic info for how to use with other vehicles.
+	Change: Upgraded the usb-serial-for-android library to latest source (from 2013-03-13). May improve compatibility with some devices. Also allows the library to be used as-is (no longer requires source code modifications). Source is still included only because no updated binary has been released yet.
+	Change: No longer displaying toast popup when app exits due to disconnect
+	Change: Updated default baud rate to most commonly used
+	Bug Fix: No longer needlessly restarts every 30 seconds when status is "monitoring" (this had been backported to master branch already)
+	Bug Fix: Fixed logic errors preventing the exit on disconnect feature from always working
+	Bug Fix: Added additional scantool initialization step to ensure buttons are recognized regardless of previous device config (turn on spaces in bus messages)

# Version 1.1 (3/17/2013)

+	Miscellaneous minor improvements, bug fixes, and project cleanup
+	New Feature: Allows launching app before connecting scantool (will request USB device permissions correctly)
+	New Feature: Ability to specify a device number to connect to in settings (allows leaving first scantool free for other apps when multiple devices are connected)
+	New Feature: Ability to specify a protocol for the scantool in settings
+	New Feature: Ability to specify the monitor command for the scantool in settings

# Version 1.0 (3/16/2013)

+	Initial public release


SteeringWheelInterface/src/com/theksmith/steeringwheelinterface/ButtonActions.java
/*
 * TODO: Re-do this class to allow end users to define the button's bus messages and map each to a predefined action
 * 
 * Currently the project is assumed to be only a template. A developer would need to override
 * this class with their own specific implementation. To make this a "real" app, we will
 * need to make the buttons and their associated actions definable via the settings screen.
 */

package com.theksmith.steeringwheelinterface;

import java.util.Date;
import java.util.HashMap;

import android.content.Context;
import android.media.AudioManager;
import android.util.Log;
import android.view.KeyEvent;


/**
 * Provides a central place to define which bus messages correspond to which buttons, and which buttons do what actions.
 * This class is a basic template that would most likely be overridden.
 * 
 * @author Kristoffer Smith <kristoffer@theksmith.com>
 */
public class ButtonActions {
	protected static final String TAG = ButtonActions.class.getSimpleName();
	
	protected static final int DEBOUNCE_THRESHOLD = 50;	//milliseconds
	
	protected Context mAppContext;
	protected HashMap<String, Long> mBusMessageDebounceTimes = new HashMap<String, Long>();
	
    //performAction() return status
	public static final int STATUS_ERROR_UNKNOWN = 0;
	public static final int STATUS_ERROR_UNKNOWNBUTTON = 1;
	public static final int STATUS_ERROR_ACTIONERROR = 2;
	public static final int STATUS_SUCCESS = 3;	

	//the known buttons and the beginning of their corresponding bus messages
	//TODO: allow wildcard or regex definitions
	public static final String BUTTON_LEFT_CENTER = "3D 11 00 80";
	public static final String BUTTON_LEFT_DOWN = "3D 11 10 00";
	public static final String BUTTON_LEFT_UP = "3D 11 20 00";
	public static final String BUTTON_RIGHT_CENTER = "3D 11 00 02";
	public static final String BUTTON_RIGHT_DOWN = "3D 11 02 00";
	public static final String BUTTON_RIGHT_UP = "3D 11 04 00";
	
	
	private ButtonActions() { 
		//exists only to prevent creation of class without passing required param
	}

	
	/**
	 * Constructor.
	 * 
	 * @param appContext		The application context of the container app. 
	 */
	public ButtonActions(Context appContext) {
		mAppContext = appContext.getApplicationContext();
	}
	

	/**
	 * Executes a particular button's assigned action.
	 * 
	 * @param forBusMessage		A bus message, expected to correspond to a button (one of the ButtonActions.BUTTON_XYZ definitions).
	 * @return 					Returns one of the ButtonActions.STATUS_XYZ definitions.
	 */
	public int performAction(String forBusMessage) {
		try {			
			if (forBusMessage.startsWith(BUTTON_LEFT_CENTER)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnMediaPause();
				}
			} else if (forBusMessage.startsWith(BUTTON_LEFT_DOWN)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnMediaTrackPrevious();
				}
			} else if (forBusMessage.startsWith(BUTTON_LEFT_UP)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnMediaTrackNext();
				}
			} else if (forBusMessage.startsWith(BUTTON_RIGHT_CENTER)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnHomeScreen();
				}
			} else if (forBusMessage.startsWith(BUTTON_RIGHT_DOWN)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnVolumeDown();
				}
			} else if (forBusMessage.startsWith(BUTTON_RIGHT_UP)) {
				if (!isHardwareBounce(forBusMessage)) {
					btnVolumeUp();
				}
			} else {
				Log.i(TAG, "Unknown button: " + forBusMessage);
				return STATUS_ERROR_UNKNOWNBUTTON;
			}
		} catch (Exception ex) {
			Log.e(TAG, "Error performing action for button: " + forBusMessage, ex);
			return STATUS_ERROR_ACTIONERROR;
		}
		
		return STATUS_SUCCESS;
	}
	

	protected boolean isHardwareBounce(String busMessage) {
		Boolean isBounce = true;		
		
		long now = (new Date()).getTime();		
		Long last = mBusMessageDebounceTimes.get(busMessage);
		
		if (last == null) {
			isBounce = false;
		} else {		
			if (now - last > DEBOUNCE_THRESHOLD) {
				isBounce = false;
			}
		}
		
		mBusMessageDebounceTimes.put(busMessage, now);
		
		return isBounce;
	}
		
	
	//TODO: most of these require root, is there a better way? how does Tasker do it?
	

	protected void btnMediaPause() throws Exception {
		Runtime.getRuntime().exec("su -c input keyevent " + KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE);
    }
    

	protected void btnMediaTrackNext() throws Exception {
   		Runtime.getRuntime().exec("su -c input keyevent " + KeyEvent.KEYCODE_MEDIA_NEXT);
    }
    
    
	protected void btnMediaTrackPrevious() throws Exception {
   		Runtime.getRuntime().exec("su -c input keyevent " +  + KeyEvent.KEYCODE_MEDIA_PREVIOUS);
    }

	
	protected void btnHomeScreen() throws Exception {
		Runtime.getRuntime().exec("su -c input keyevent " + KeyEvent.KEYCODE_HOME);
    }

    
	protected void btnVolumeUp() throws Exception {
   		AudioManager mAudioManager = (AudioManager)mAppContext.getSystemService(Context.AUDIO_SERVICE);    		
   		mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_RAISE, AudioManager.FLAG_SHOW_UI);
    }
    
    
	protected void btnVolumeDown() throws Exception {
		AudioManager mAudioManager = (AudioManager)mAppContext.getSystemService(Context.AUDIO_SERVICE);    		
		mAudioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_LOWER, AudioManager.FLAG_SHOW_UI);
    }    
}

usb-serial-for-android/UsbSerialLibrary/src/com/hoho/android/usbserial/driver/UsbSerialProber.java
/* Copyright 2011 Google Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * Project home page: http://code.google.com/p/usb-serial-for-android/
 */

package com.hoho.android.usbserial.driver;

import java.util.Map;

import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbManager;

/**
 * Helper class to assist in detecting and building {@link UsbSerialDriver}
 * instances from available hardware.
 *
 * @author mike wakerly (opensource@hoho.com)
 */
public enum UsbSerialProber {

    // TODO(mikey): Too much boilerplate.

    /**
     * Prober for {@link FtdiSerialDriver}.
     *
     * @see FtdiSerialDriver
     */
    FTDI_SERIAL {
        @Override
        public UsbSerialDriver getDevice(final UsbManager manager, final UsbDevice usbDevice) {
            if (!testIfSupported(usbDevice, FtdiSerialDriver.getSupportedDevices())) {
                return null;
            }
            final UsbDeviceConnection connection = manager.openDevice(usbDevice);
            if (connection == null) {
                return null;
            }
            return new FtdiSerialDriver(usbDevice, connection);
        }
    },

    CDC_ACM_SERIAL {
        @Override
        public UsbSerialDriver getDevice(UsbManager manager, UsbDevice usbDevice) {
            if (!testIfSupported(usbDevice, CdcAcmSerialDriver.getSupportedDevices())) {
               return null;
            }
            final UsbDeviceConnection connection = manager.openDevice(usbDevice);
            if (connection == null) {
                return null;
            }
            return new CdcAcmSerialDriver(usbDevice, connection);
        }
    };

    /**
     * Builds a new {@link UsbSerialDriver} instance from the raw device, or
     * returns <code>null</code> if it could not be built (for example, if the
     * probe failed).
     *
     * @param manager the {@link UsbManager} to use
     * @param usbDevice the raw {@link UsbDevice} to use
     * @return the first available {@link UsbSerialDriver}, or {@code null} if
     *         no devices could be acquired
     */
    public abstract UsbSerialDriver getDevice(final UsbManager manager, final UsbDevice usbDevice);

    /**
     * Acquires and returns the first available serial device among all
     * available {@link UsbDevice}s, or returns {@code null} if no device could
     * be acquired.
     *
     * @param usbManager the {@link UsbManager} to use
     * @return the first available {@link UsbSerialDriver}, or {@code null} if
     *         no devices could be acquired
     */
    public static UsbSerialDriver acquire(final UsbManager usbManager) {
        for (final UsbDevice usbDevice : usbManager.getDeviceList().values()) {
            final UsbSerialDriver probedDevice = acquire(usbManager, usbDevice);
            if (probedDevice != null) {
                return probedDevice;
            }
        }
        return null;
    }

    /**
     * Builds and returns a new {@link UsbSerialDriver} from the given
     * {@link UsbDevice}, or returns {@code null} if no drivers supported this
     * device.
     *
     * @param usbManager the {@link UsbManager} to use
     * @param usbDevice the {@link UsbDevice} to use
     * @return a new {@link UsbSerialDriver}, or {@code null} if no devices
     *         could be acquired
     */
    public static UsbSerialDriver acquire(final UsbManager usbManager, final UsbDevice usbDevice) {
        for (final UsbSerialProber prober : values()) {
            final UsbSerialDriver probedDevice = prober.getDevice(usbManager, usbDevice);
            if (probedDevice != null) {
                return probedDevice;
            }
        }
        return null;
    }

    /**
     * Returns {@code true} if the given device is found in the vendor/product map.
     *
     * @param usbDevice the device to test
     * @param supportedDevices map of vendor ids to product id(s)
     * @return {@code true} if supported
     */
    public static boolean testIfSupported(final UsbDevice usbDevice,
            final Map<Integer, int[]> supportedDevices) {        
        /* Kristoffer Smith
         * changed method to public to support acquiring a specific device from consuming code
         */
        
        final int[] supportedProducts = supportedDevices.get(
                Integer.valueOf(usbDevice.getVendorId()));
        if (supportedProducts == null) {
            return false;
        }

        final int productId = usbDevice.getProductId();
        for (int supportedProductId : supportedProducts) {
            if (productId == supportedProductId) {
                return true;
            }
        }
        return false;
    }

}


usb-serial-for-android/UsbSerialLibrary/src/com/hoho/android/usbserial/driver/FtdiSerialDriver.java
/* Copyright 2011 Google Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 * USA.
 *
 * Project home page: http://code.google.com/p/usb-serial-for-android/
 */

package com.hoho.android.usbserial.driver;

import android.hardware.usb.UsbConstants;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbRequest;
import android.util.Log;

import com.hoho.android.usbserial.util.HexDump;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * A {@link UsbSerialDriver} implementation for a variety of FTDI devices
 * <p>
 * This driver is based on
 * <a href="http://www.intra2net.com/en/developer/libftdi">libftdi</a>, and is
 * copyright and subject to the following terms:
 *
 * <pre>
 *   Copyright (C) 2003 by Intra2net AG
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License
 *   version 2.1 as published by the Free Software Foundation;
 *
 *   opensource@intra2net.com
 *   http://www.intra2net.com/en/developer/libftdi
 * </pre>
 *
 * </p>
 * <p>
 * Some FTDI devices have not been tested; see later listing of supported and
 * unsupported devices. Devices listed as "supported" support the following
 * features:
 * <ul>
 * <li>Read and write of serial data (see {@link #read(byte[], int)} and
 * {@link #write(byte[], int)}.
 * <li>Setting baud rate (see {@link #setBaudRate(int)}).
 * </ul>
 * </p>
 * <p>
 * Supported and tested devices:
 * <ul>
 * <li>{@value DeviceType#TYPE_R}</li>
 * </ul>
 * </p>
 * <p>
 * Unsupported but possibly working devices (please contact the author with
 * feedback or patches):
 * <ul>
 * <li>{@value DeviceType#TYPE_2232C}</li>
 * <li>{@value DeviceType#TYPE_2232H}</li>
 * <li>{@value DeviceType#TYPE_4232H}</li>
 * <li>{@value DeviceType#TYPE_AM}</li>
 * <li>{@value DeviceType#TYPE_BM}</li>
 * </ul>
 * </p>
 *
 * @author mike wakerly (opensource@hoho.com)
 * @see <a href="http://code.google.com/p/usb-serial-for-android/">USB Serial
 * for Android project page</a>
 * @see <a href="http://www.ftdichip.com/">FTDI Homepage</a>
 * @see <a href="http://www.intra2net.com/en/developer/libftdi">libftdi</a>
 */
public class FtdiSerialDriver extends UsbSerialDriver {

    private static final int DEFAULT_BAUD_RATE = 115200;
    private static final int DEFAULT_DATA_BITS = DATABITS_8;
    private static final int DEFAULT_PARITY = PARITY_NONE;
    private static final int DEFAULT_STOP_BITS = STOPBITS_1;

    public static final int USB_TYPE_STANDARD = 0x00 << 5;
    public static final int USB_TYPE_CLASS = 0x00 << 5;
    public static final int USB_TYPE_VENDOR = 0x00 << 5;
    public static final int USB_TYPE_RESERVED = 0x00 << 5;

    public static final int USB_RECIP_DEVICE = 0x00;
    public static final int USB_RECIP_INTERFACE = 0x01;
    public static final int USB_RECIP_ENDPOINT = 0x02;
    public static final int USB_RECIP_OTHER = 0x03;

    public static final int USB_ENDPOINT_IN = 0x80;
    public static final int USB_ENDPOINT_OUT = 0x00;

    public static final int USB_WRITE_TIMEOUT_MILLIS = 5000;
    public static final int USB_READ_TIMEOUT_MILLIS = 5000;

    // From ftdi.h
    /**
     * Reset the port.
     */
    private static final int SIO_RESET_REQUEST = 0;

    /**
     * Set the modem control register.
     */
    private static final int SIO_MODEM_CTRL_REQUEST = 1;

    /**
     * Set flow control register.
     */
    private static final int SIO_SET_FLOW_CTRL_REQUEST = 2;

    /**
     * Set baud rate.
     */
    private static final int SIO_SET_BAUD_RATE_REQUEST = 3;

    /**
     * Set the data characteristics of the port.
     */
    private static final int SIO_SET_DATA_REQUEST = 4;

    private static final int SIO_RESET_SIO = 0;

    public static final int FTDI_DEVICE_OUT_REQTYPE =
            UsbConstants.USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_OUT;

    public static final int FTDI_DEVICE_IN_REQTYPE =
            UsbConstants.USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN;

    /**
     * Length of the modem status header, transmitted with every read.
     */
    private static final int MODEM_STATUS_HEADER_LENGTH = 2;

    private final String TAG = FtdiSerialDriver.class.getSimpleName();

    private DeviceType mType;

    /**
     * FTDI chip types.
     */
    private static enum DeviceType {
        TYPE_BM, TYPE_AM, TYPE_2232C, TYPE_R, TYPE_2232H, TYPE_4232H;
    }

    private int mInterface = 0; /* INTERFACE_ANY */

    private int mMaxPacketSize = 64; // TODO(mikey): detect

    //KLS: changed initial values so we can detect if these are set by the user or if we should use defaults
    private int mBaudRate = -1;
    private int mDataBits = -1;
    private int mParity = -1;
    private int mStopBits = -1;

    /**
     * Due to http://b.android.com/28023 , we cannot use UsbRequest async reads
     * since it gives no indication of number of bytes read. Set this to
     * {@code true} on platforms where it is fixed.
     */
    private static final boolean ENABLE_ASYNC_READS = false;

    /**
     * Constructor.
     *
     * @param usbDevice the {@link UsbDevice} to use
     * @param usbConnection the {@link UsbDeviceConnection} to use
     * @throws UsbSerialRuntimeException if the given device is incompatible
     *             with this driver
     */
    public FtdiSerialDriver(UsbDevice usbDevice, UsbDeviceConnection usbConnection) {
        super(usbDevice, usbConnection);
        mType = null;
    }

    public void reset() throws IOException {
        int result = mConnection.controlTransfer(FTDI_DEVICE_OUT_REQTYPE, SIO_RESET_REQUEST,
                SIO_RESET_SIO, 0 /* index */, null, 0, USB_WRITE_TIMEOUT_MILLIS);
        if (result != 0) {
            throw new IOException("Reset failed: result=" + result);
        }

        // TODO(mikey): autodetect.
        mType = DeviceType.TYPE_R;
    }

    @Override
    public void open() throws IOException {
        boolean opened = false;
        try {
            for (int i = 0; i < mDevice.getInterfaceCount(); i++) {
                if (mConnection.claimInterface(mDevice.getInterface(i), true)) {
                    Log.d(TAG, "claimInterface " + i + " SUCCESS");
                } else {
                    Log.d(TAG, "claimInterface " + i + " FAIL");
                }
            }
            reset();
            
            /* Kristoffer Smith 
             * changed to use default for each parameter ONLY when not already set by user 
             */
            if (mBaudRate <= 0) mBaudRate = DEFAULT_BAUD_RATE;
            if (mDataBits <= 0) mDataBits = DEFAULT_DATA_BITS;
            if (mStopBits <= 0) mStopBits = DEFAULT_STOP_BITS;
            if (mParity <= 0) mParity = DEFAULT_PARITY;
            setParameters(mBaudRate, mDataBits, mStopBits, mParity);
            
            opened = true;
        } finally {
            if (!opened) {
                close();
            }
        }
    }

    @Override
    public void close() {
        mConnection.close();
    }

    @Override
    public int read(byte[] dest, int timeoutMillis) throws IOException {
        final UsbEndpoint endpoint = mDevice.getInterface(0).getEndpoint(0);

        if (ENABLE_ASYNC_READS) {
            final int readAmt;
            synchronized (mReadBufferLock) {
                // mReadBuffer is only used for maximum read size.
                readAmt = Math.min(dest.length, mReadBuffer.length);
            }

            final UsbRequest request = new UsbRequest();
            request.initialize(mConnection, endpoint);

            final ByteBuffer buf = ByteBuffer.wrap(dest);
            if (!request.queue(buf, readAmt)) {
                throw new IOException("Error queueing request.");
            }

            final UsbRequest response = mConnection.requestWait();
            if (response == null) {
                throw new IOException("Null response");
            }

            final int payloadBytesRead = buf.position() - MODEM_STATUS_HEADER_LENGTH;
            if (payloadBytesRead > 0) {
                Log.d(TAG, HexDump.dumpHexString(dest, 0, Math.min(32, dest.length)));
                return payloadBytesRead;
            } else {
                return 0;
            }
        } else {
            final int totalBytesRead;

            synchronized (mReadBufferLock) {
                final int readAmt = Math.min(dest.length, mReadBuffer.length);
                totalBytesRead = mConnection.bulkTransfer(endpoint, mReadBuffer,
                        readAmt, timeoutMillis);
            }

            if (totalBytesRead < MODEM_STATUS_HEADER_LENGTH) {
                throw new IOException("Expected at least " + MODEM_STATUS_HEADER_LENGTH + " bytes");
            }

            final int payloadBytesRead = totalBytesRead - MODEM_STATUS_HEADER_LENGTH;
            if (payloadBytesRead > 0) {
                System.arraycopy(mReadBuffer, MODEM_STATUS_HEADER_LENGTH, dest, 0, payloadBytesRead);
            }
            return payloadBytesRead;
        }
    }

    @Override
    public int write(byte[] src, int timeoutMillis) throws IOException {
        final UsbEndpoint endpoint = mDevice.getInterface(0).getEndpoint(1);
        int offset = 0;

        while (offset < src.length) {
            final int writeLength;
            final int amtWritten;

            synchronized (mWriteBufferLock) {
                final byte[] writeBuffer;

                writeLength = Math.min(src.length - offset, mWriteBuffer.length);
                if (offset == 0) {
                    writeBuffer = src;
                } else {
                    // bulkTransfer does not support offsets, make a copy.
                    System.arraycopy(src, offset, mWriteBuffer, 0, writeLength);
                    writeBuffer = mWriteBuffer;
                }

                amtWritten = mConnection.bulkTransfer(endpoint, writeBuffer, writeLength,
                        timeoutMillis);
            }

            if (amtWritten <= 0) {
                throw new IOException("Error writing " + writeLength
                        + " bytes at offset " + offset + " length=" + src.length);
            }

            Log.d(TAG, "Wrote amtWritten=" + amtWritten + " attempted=" + writeLength);
            offset += amtWritten;
        }
        return offset;
    }

    @Override
    @Deprecated
    public int setBaudRate(int baudRate) throws IOException {
        long[] vals = convertBaudrate(baudRate);
        long actualBaudrate = vals[0];
        long index = vals[1];
        long value = vals[2];
        int result = mConnection.controlTransfer(FTDI_DEVICE_OUT_REQTYPE,
                SIO_SET_BAUD_RATE_REQUEST, (int) value, (int) index,
                null, 0, USB_WRITE_TIMEOUT_MILLIS);
        if (result != 0) {
            throw new IOException("Setting baudrate failed: result=" + result);
        }
        return (int) actualBaudrate;
    }

    @Override
    public void setParameters(int baudRate, int dataBits, int stopBits, int parity)
            throws IOException {
        mBaudRate = setBaudRate(baudRate);

        int config = dataBits;

        switch (parity) {
            case PARITY_NONE:
                config |= (0x00 << 8);
                break;
            case PARITY_ODD:
                config |= (0x01 << 8);
                break;
            case PARITY_EVEN:
                config |= (0x02 << 8);
                break;
            case PARITY_MARK:
                config |= (0x03 << 8);
                break;
            case PARITY_SPACE:
                config |= (0x04 << 8);
                break;
            default:
                throw new IllegalArgumentException("Unknown parity value: " + parity);
        }

        switch (stopBits) {
            case STOPBITS_1:
                config |= (0x00 << 11);
                break;
            case STOPBITS_1_5:
                config |= (0x01 << 11);
                break;
            case STOPBITS_2:
                config |= (0x02 << 11);
                break;
            default:
                throw new IllegalArgumentException("Unknown stopBits value: " + stopBits);
        }

        int result = mConnection.controlTransfer(FTDI_DEVICE_OUT_REQTYPE,
                SIO_SET_DATA_REQUEST, config, 0 /* index */,
                null, 0, USB_WRITE_TIMEOUT_MILLIS);
        if (result != 0) {
            throw new IOException("Setting parameters failed: result=" + result);
        }

        mParity = parity;
        mStopBits = stopBits;
        mDataBits = dataBits;
    }

    private long[] convertBaudrate(int baudrate) {
        // TODO(mikey): Braindead transcription of libfti method.  Clean up,
        // using more idiomatic Java where possible.
        int divisor = 24000000 / baudrate;
        int bestDivisor = 0;
        int bestBaud = 0;
        int bestBaudDiff = 0;
        int fracCode[] = {
                0, 3, 2, 4, 1, 5, 6, 7
        };

        for (int i = 0; i < 2; i++) {
            int tryDivisor = divisor + i;
            int baudEstimate;
            int baudDiff;

            if (tryDivisor <= 8) {
                // Round up to minimum supported divisor
                tryDivisor = 8;
            } else if (mType != DeviceType.TYPE_AM && tryDivisor < 12) {
                // BM doesn't support divisors 9 through 11 inclusive
                tryDivisor = 12;
            } else if (divisor < 16) {
                // AM doesn't support divisors 9 through 15 inclusive
                tryDivisor = 16;
            } else {
                if (mType == DeviceType.TYPE_AM) {
                    // TODO
                } else {
                    if (tryDivisor > 0x1FFFF) {
                        // Round down to maximum supported divisor value (for
                        // BM)
                        tryDivisor = 0x1FFFF;
                    }
                }
            }

            // Get estimated baud rate (to nearest integer)
            baudEstimate = (24000000 + (tryDivisor / 2)) / tryDivisor;

            // Get absolute difference from requested baud rate
            if (baudEstimate < baudrate) {
                baudDiff = baudrate - baudEstimate;
            } else {
                baudDiff = baudEstimate - baudrate;
            }

            if (i == 0 || baudDiff < bestBaudDiff) {
                // Closest to requested baud rate so far
                bestDivisor = tryDivisor;
                bestBaud = baudEstimate;
                bestBaudDiff = baudDiff;
                if (baudDiff == 0) {
                    // Spot on! No point trying
                    break;
                }
            }
        }

        // Encode the best divisor value
        long encodedDivisor = (bestDivisor >> 3) | (fracCode[bestDivisor & 7] << 14);
        // Deal with special cases for encoded value
        if (encodedDivisor == 1) {
            encodedDivisor = 0; // 3000000 baud
        } else if (encodedDivisor == 0x4001) {
            encodedDivisor = 1; // 2000000 baud (BM only)
        }

        // Split into "value" and "index" values
        long value = encodedDivisor & 0xFFFF;
        long index;
        if (mType == DeviceType.TYPE_2232C || mType == DeviceType.TYPE_2232H
                || mType == DeviceType.TYPE_4232H) {
            index = (encodedDivisor >> 8) & 0xffff;
            index &= 0xFF00;
            index |= 0 /* TODO mIndex */;
        } else {
            index = (encodedDivisor >> 16) & 0xffff;
        }

        // Return the nearest baud rate
        return new long[] {
                bestBaud, index, value
        };
    }

    @Override
    public boolean getCD() throws IOException {
        return false;
    }

    @Override
    public boolean getCTS() throws IOException {
        return false;
    }

    @Override
    public boolean getDSR() throws IOException {
        return false;
    }

    @Override
    public boolean getDTR() throws IOException {
        return false;
    }

    @Override
    public void setDTR(boolean value) throws IOException {
    }

    @Override
    public boolean getRI() throws IOException {
        return false;
    }

    @Override
    public boolean getRTS() throws IOException {
        return false;
    }

    @Override
    public void setRTS(boolean value) throws IOException {
    }

    public static Map<Integer, int[]> getSupportedDevices() {
        final Map<Integer, int[]> supportedDevices = new LinkedHashMap<Integer, int[]>();
        supportedDevices.put(Integer.valueOf(UsbId.VENDOR_FTDI),
                new int[] {
                    UsbId.FTDI_FT232R,
                });
        return supportedDevices;
    }

}
